lib/elixir/lib/calendar/iso.ex

```diff
@@ -1455,6 +1455,229 @@ defmodule Calendar.ISO do
     {days, {@parts_per_day - 1, @parts_per_day}}
   end
 
+  @doc """
+  Shift Date by Duration according to its calendar.
+
+  ## Examples
+
+      iex> Calendar.ISO.shift_date(2016, 1, 3, Duration.new(month: 2))
+      {2016, 3, 3}
+      iex> Calendar.ISO.shift_date(2016, 2, 29, Duration.new(month: 1))
+      {2016, 3, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, Duration.new(month: 1))
+      {2016, 2, 29}
+      iex> Calendar.ISO.shift_date(2016, 1, 31, Duration.new(year: 4, day: 1))
+      {2020, 2, 1}
+  """
+  @spec shift_date(year, month, day, Duration.t()) :: {year, month, day}
+  @impl true
+  def shift_date(year, month, day, duration) do
+    validate_shift_fields!(duration, :date)
+
+    shift_options = get_shift_options(:date, duration)
```

Comment:
Nitpick: let's merge these two! You can pattern match on `hour: 0, minute: 0, second: 0, microsecond: {0, 0}` in `get_shift_options` and then have a clause that finds the one that is non-zero in case of errors. This makes the happy path super fast!

Also, I think we should call it `get_shift_date_options`. That's because the different `get_shift_options` clauses really share nothing.